### PR TITLE
atrous: cut memory use in half

### PR DESCRIPTION
**Pull request by ralfbrown**
Instead of collecting all of the detail scales produced during the wavelet decomposition and then synthesizing them together in a separate pass, reversing the order of synthesis allows us to reuse the same buffer for each detail scale.  This means that the number of buffers needed drops from one scratch buffer plus 8 scale buffers (for images >= 1024px on the short side) to just two scratch buffers plus the single scale buffer.  In the terms used by tiling_callback(), tiling->factor decreases from 11 to 5.

This can dramatically speed up processing under low-memory conditions.  Using mire1-xtrans.raf and the old default host_memory_limit of 1500, wall time with 32 threads decreases from 2.7 seconds to 1.2 second as the number of tiles drops from 35 to 5.

Even with enough memory to avoid tiling, there is still a slight performance improvement (measured ~3.5% speedup using 4, 8, and 32 threads).

The altered order of execution does change the output slightly (max dE 1.5779, 159 pixels changed, PAE the smallest representable difference).

**Someone who knows OpenCL should apply the same reordering to the GPU code path to let the contrast equalizer run on cards with less RAM.**

### CODE REVIEW

**Comment by ralfbrown** [comment]
BTW, additional optimizations should be possible with the reordered computation.  I think I can eliminate the detail-scale buffer....

**Comment by TurboGit** [comment]
> BTW, additional optimizations should be possible with the reordered computation. I think I can eliminate the detail-scale buffer....

Ok, I'll wait for this before merging. Thanks.

**Comment by ralfbrown** [comment]
The reason for holding off on a merge is less that there may be more optimizations coming than that the current code won't actually help the tiling because `tiling_callback` still has to report the much larger memory requirement of the OpenCL path.  Two solutions to that:

- make the same revision in the OpenCL path.  I have stubbed out a suggested version which only needs two trivial OpenCL kernels to be written and a bit of testing (famous last words?)
- allow `tiling_callback` to return separate memory sizes for CPU and OpenCL.  This is something that would be useful regardless, but I'm not looking forward to trying it because it will involve touching critical code in a path I can't test.

**Comment by ralfbrown** [comment]
Rebased to current master, and now reports separate memory requirements for CPU and OpenCL.
Don't wait on me to apply further optimizations -- going from tiling->factor 5.0 to 4.0 will involve bigger and trickier changes than going from 11 to 5 did, and my spare time is about to evaporate.

File: src/iop/atrous.c
Comment (TurboGit) [inline src/iop/atrous.c:261]: this call seems out of sync with the imagebuf definition:

```
gboolean dt_iop_alloc_image_buffers(struct dt_iop_module_t *const module,
                                    const struct dt_iop_roi_t *const roi_in,
                                    const struct dt_iop_roi_t *const roi_out, ...);
```

The NULL before roi_in is suspicious!

**Review by TurboGit — CHANGES_REQUESTED** [review]
API not in sync.

**Comment by TurboGit** [comment]
Working ok on OpenCL but crash on integration test 0024-contrast-equalizer probably due to error reported above.

File: src/iop/atrous.c
Comment (ralfbrown) [reply src/iop/atrous.c]: Oops, missed that one because it didn't generate a merge conflict.

**Comment by TurboGit** [comment]
Thanks for the record, the dE goes from 0 to 1.5:

```
Test 0024-contrast-equalizer
      Image mire1.cr2
      CPU & GPU version differ by 784006 pixels
      Max  dE         1.5779
  OK
```

**Pull request merged**
